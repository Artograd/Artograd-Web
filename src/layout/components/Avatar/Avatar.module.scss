.avatarMenu {
  margin: 0 12px;
}

.name,
.avatarMenuItem {
  font-weight: 400;
  font-size: 14px;
  line-height: 24px;
  color: var(--night800);
}

.avatarMenuItem svg {
  fill: var(--night600);
}

.email {
  font-weight: 400;
  font-size: 14px;
  line-height: 24px;
  color: var(--night600);
}
