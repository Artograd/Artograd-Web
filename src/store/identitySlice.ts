import { createSlice } from '@reduxjs/toolkit';
import storage from 'redux-persist/lib/storage';
import { persistReducer } from 'redux-persist';

export type IdentityStateType = {
  'cognito:username': string;
  email: string;
  email_verified: boolean;
  family_name: string;
  gender: 'male' | 'female' | '';
  given_name: string;
  phone_number: string;
  phone_number_verified: boolean;
  isLoggedIn: boolean;
};

export type InitialStateType = {
  identity: IdentityStateType;
};

export const identityState: IdentityStateType = {
  'cognito:username': '',
  email: '',
  email_verified: false,
  family_name: '',
  gender: '',
  given_name: '',
  phone_number: '',
  phone_number_verified: false,
  isLoggedIn: false,
};

export const initialState: InitialStateType = {
  identity: identityState,
};

export const identitySlice = createSlice({
  name: 'identity',
  initialState,
  reducers: {
    saveUserData: (state, action) => {
      state.identity = action.payload;
    },
    userLogin: (state, action) => {
      state.identity.isLoggedIn = action.payload;
    },
  },
});

export const { saveUserData, userLogin } = identitySlice.actions;

export default identitySlice.reducer;

const persistConfig = {
  key: 'identity',
  version: 1,
  storage,
};

export const persistedIdentityReducer = persistReducer(
  persistConfig,
  identitySlice.reducer,
);
