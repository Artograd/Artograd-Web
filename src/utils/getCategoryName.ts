import { CategoryItemType } from '../types';

export const categoryList: CategoryItemType[] = [
  { id: 0, name: 'tendersPage.newTender.categories.sculptures' },
  { id: 1, name: 'tendersPage.newTender.categories.mosaics' },
  { id: 2, name: 'tendersPage.newTender.categories.murals' },
  { id: 3, name: 'tendersPage.newTender.categories.graffiti' },
  { id: 4, name: 'tendersPage.newTender.categories.functionalArt' },
  {
    id: 5,
    name: 'tendersPage.newTender.categories.interactiveInstallations',
  },
  { id: 6, name: 'tendersPage.newTender.categories.botanicalArt' },
  { id: 7, name: 'tendersPage.newTender.categories.waterFeatures' },
  { id: 8, name: 'tendersPage.newTender.categories.themedGardens' },
  { id: 9, name: 'tendersPage.newTender.categories.recycled' },
];

export const getCategoryName = (id: string) => {
  return categoryList.find((category) => category.id === Number(id));
};
