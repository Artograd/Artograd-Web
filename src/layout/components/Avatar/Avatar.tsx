import {
  DropdownMenuBody,
  DropdownMenuButton,
  MainMenuAvatar,
  DropdownMenuSplitter,
  DropdownMenuHeader,
  Dropdown,
} from '@epam/uui';
import { ReactComponent as ProfileIcon } from '@epam/assets/icons/common/social-person-18.svg';
import { ReactComponent as SettingsIcon } from '@epam/assets/icons/common/action-settings-18.svg';
import { ReactComponent as BellIcon } from '@epam/assets/icons/common/bell-18.svg';
import { ReactComponent as LogoutIcon } from '@epam/assets/icons/common/navigation-logout-18.svg';
import { identityState, saveUserData } from '../../../store/identitySlice';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '../../../store/store';
import { useHistory } from 'react-router-dom';
import styles from './Avatar.module.scss';
import { useTranslation } from 'react-i18next';

export const Avatar = () => {
  const history = useHistory();
  const dispatch = useDispatch();
  const { t } = useTranslation();
  const { family_name, name, email } = useSelector(
    (state: RootState) => state.identity,
  );
  const logOut = async () => {
    const refreshToken = localStorage.getItem('refresh_token');
    // clear identity state to the initial
    dispatch(saveUserData(identityState));
    // revoke cognito token and clear tokens from localStorage
    await fetch(
      `${process.env.REACT_APP_COGNITO_URL}/oauth2/revoke?token=${refreshToken}&client_id=${process.env.REACT_APP_CLIENT_ID}`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
      },
    )
      .then(() => {
        localStorage.removeItem('id_token');
        localStorage.removeItem('refresh_token');
        localStorage.removeItem('access_token');
      })
      .finally(() => history.push('/'));
  };

  return (
    <Dropdown
      key="avatar"
      renderTarget={(props) => (
        <MainMenuAvatar
          avatarUrl="https://api.dicebear.com/7.x/pixel-art/svg?seed=Coco&radius=50&backgroundColor=b6e3f4"
          isDropdown
          {...props}
        />
      )}
      renderBody={(props) => (
        <DropdownMenuBody {...props} cx={styles.userMenu}>
          <DropdownMenuHeader
            caption={`${name} ${family_name}`}
            cx={styles.name}
          />
          <DropdownMenuHeader caption={email} cx={styles.email} />
          <DropdownMenuSplitter />
          <DropdownMenuButton
            caption={t('global.layout.header.profile')}
            cx={styles.userMenuItem}
            icon={ProfileIcon}
            onClick={() => history.push('/profile')}
          />
          <DropdownMenuButton
            caption={t('global.layout.header.settings')}
            cx={styles.userMenuItem}
            icon={SettingsIcon}
            onClick={() => history.push('/settings')}
          />
          <DropdownMenuButton
            caption={t('global.layout.header.notifications')}
            cx={styles.userMenuItem}
            icon={BellIcon}
            onClick={() => history.push('/notifications')}
          />
          <DropdownMenuSplitter />
          <DropdownMenuButton
            caption={t('global.layout.header.logOut')}
            cx={styles.userMenuItem}
            icon={LogoutIcon}
            onClick={() => logOut()}
          />
        </DropdownMenuBody>
      )}
      placement="bottom-end"
    />
  );
};
