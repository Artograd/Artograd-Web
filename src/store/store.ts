import { configureStore } from '@reduxjs/toolkit';
import {
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
  persistCombineReducers,
} from 'redux-persist';
import { persistedIdentityReducer } from './identitySlice';
import { persistedProfileOverviewReducer } from './slices/profileOverviewSlice';
import storage from 'redux-persist/lib/storage';
import { persistedProfileInformationReducer } from './slices/ProfileInformationSlice';

const persistConfig = {
  key: 'root',
  storage,
};

const reducer = persistCombineReducers(persistConfig, {
  persistedIdentityReducer,
  persistedProfileOverviewReducer,
  persistedProfileInformationReducer,
});

export const store = configureStore({
  reducer: reducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
