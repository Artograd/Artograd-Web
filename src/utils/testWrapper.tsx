import { renderWithContextAsync } from '@epam/uui-test-utils';
import { MemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import { store } from '../store/store';

export const testWrapper = async ({
  component,
  history,
}: {
  component: React.JSX.Element;
  history: MemoryHistory<unknown>;
  
}) => {
  return renderWithContextAsync(<Provider store={store}><Router history={history}>{component}</Router></Provider>);
};
