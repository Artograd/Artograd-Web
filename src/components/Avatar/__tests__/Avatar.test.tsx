import { Avatar } from '../Avatar';
import { createMemoryHistory } from 'history';
import { testWrapper } from '../../../utils/testWrapper';
import { identityState } from '../../../store/identitySlice';
import { act, fireEvent, screen, userEvent } from '@epam/uui-test-utils';

describe('Avatar', () => {
  const history = createMemoryHistory();
  afterEach(() => {
    jest.clearAllMocks();
  });
  test('renders correctly', async () => {
    const user = userEvent.setup();

    const component = await testWrapper({
      component: <Avatar />,
      history,
      state: {
        identity: {
          ...identityState,
          given_name: 'test',
          family_name: 'user',
          email: 'email@email.com',
        },
      },
    });
    await act(async () => {
      user.click(screen.getByTestId('user-avatar'));

      expect(component).toMatchSnapshot();
    });
  });

  test('should logout on menu item click', async () => {
    jest.spyOn(Storage.prototype, 'removeItem');

    act(async () => {
      await testWrapper({
        component: <Avatar />,
        history,
        state: {
          identity: {
            ...identityState,
            given_name: 'test',
            family_name: 'user',
            email: 'email@email.com',
          },
        },
      });

      fireEvent.click(screen.getByTestId('user-avatar'));
      fireEvent.click(screen.getByText('Log out'));

      expect(localStorage.removeItem).toBeCalledTimes(3);
    });
  });
});
