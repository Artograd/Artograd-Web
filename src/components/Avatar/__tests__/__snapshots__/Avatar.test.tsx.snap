// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="uui-panel nNCe-- uEZmHy"
        data-testid="user-avatar"
      >
        <button
          class="-Ar0uP -clickable"
        >
          <img
            class="hd78KB"
            height="36"
            src="https://static.cdn.epam.com/uploads/690afa39a93c88c4dd13758fe1d869d5/EPM-UUI/Images/avatar_placeholder.jpg"
            width="36"
          />
          <div>
            <div
              class="Prmk3w uui-icon uui-enabled"
            >
              <svg
                class=""
                height="18"
                viewBox="0 0 18 18"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12.443 6.442 9 9.877 5.558 6.442 4.5 7.5 9 12l4.5-4.5z"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </div>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="uui-panel nNCe-- uEZmHy"
      data-testid="user-avatar"
    >
      <button
        class="-Ar0uP -clickable"
      >
        <img
          class="hd78KB"
          height="36"
          src="https://static.cdn.epam.com/uploads/690afa39a93c88c4dd13758fe1d869d5/EPM-UUI/Images/avatar_placeholder.jpg"
          width="36"
        />
        <div>
          <div
            class="Prmk3w uui-icon uui-enabled"
          >
            <svg
              class=""
              height="18"
              viewBox="0 0 18 18"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12.443 6.442 9 9.877 5.558 6.442 4.5 7.5 9 12l4.5-4.5z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </div>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByAria": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByRoleAndText": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByAria": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByRoleAndText": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByAria": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByRoleAndText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByAria": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByRoleAndText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByAria": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByRoleAndText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByAria": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByRoleAndText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
